src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -666,77 +694,127 @@ def same_range(self, index: int, array2, index2: int) -> bool:
+        if not isinstance(array2.indices[index2], Range):
             raise TypeError(
                 f"The child of the second array argument at the specified "
                 f"index '{index2}' should be a Range node, but found "
-                f"'{type(array2.children[index2]).__name__}'.")
+                f"'{type(array2.indices[index2]).__name__}'.")
 
-        range1 = self.children[index]
-        range2 = array2.children[index2]
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.assignment import Assignment
+        range1 = self.indices[index]
+        range2 = array2.indices[index2]
 
         sym_maths = SymbolicMaths.get()
-        # compare lower bounds
-        if self.is_lower_bound(index) and array2.is_lower_bound(index2):
-            # Both self and array2 use the lbound() intrinsic to
-            # specify the lower bound of the array dimension. We may
-            # not be able to determine what the lower bounds of these
-            # arrays are statically but at runtime the code will fail
-            # if the ranges do not match so we assume that the lower
-            # bounds are consistent.
-            pass
-        elif self.is_lower_bound(index) or array2.is_lower_bound(index2):
-            # One and only one of self and array2 use the lbound()
-            # intrinsic to specify the lower bound of the array
-            # dimension. In this case assume that the ranges are
-            # different (although they could potentially be the same).
-            return False
-        elif not sym_maths.equal(range1.start, range2.start):
-            # Neither self nor array2 use the lbound() intrinsic to
-            # specify the lower bound of the array dimension. Try to
-            # determine if they are the same by matching the
-            # text. Use symbolic maths to do the comparison.
-            return False
+        # The logic below assumes array expressions come from valid Fortran,
+        # (e.g. a(2:4) = b(2:5) is not valid Fortran)
+        # and therefore, we assume the length of the equivalent ranges in the
+        # same statement matches:
+        # e.g. a(3, :) and b(4:, 4)
+        # a dim 2 and b dim 1 must have the same length, but not necessarely
+        # the same range (lower and upper bounds)
+        n_range1 = len([x for x in self.children[:index]
+                        if isinstance(x, Range)])
+        n_range2 = len([x for x in array2.children[:index2]
+                        if isinstance(x, Range)])
+        assume_same_length = (
+            (self.ancestor(Assignment) is array2.ancestor(Assignment) or
+             self.is_same_array(array2)) and n_range1 == n_range2
+        )
+
+        # Try to get the datatypes, only if they are ArrayType (so we know
+        # that these have the shape attribute)
+        array1_type = None
+        if (isinstance(self, Reference) and isinstance(self.symbol, DataSymbol)
```

Comment:
I am trying to avoid the reference.is_array because its meaning is ambiguous (is the datatype of the symbol or of the subtree?). I plan to address this in #2608